bt-core/src/main/java/bt/data/digest/SHA1Digester.java

```diff
@@ -17,6 +17,16 @@
 package bt.data.digest;
 
 public class SHA1Digester extends JavaSecurityDigester {
+    private static final int DEFAULT_STEP_SIZE = 1 << 15; // 32KiB
```

Comment:
Why so low? I recall starting in the same ballpark and then having to increase this parameter significantly to improve the performance.

Reply:
That's really interesting. I haven't performance tested this code. I will do so, and get back to you with how this buffer size affects runtime.

I have used digest APIs in the past and gotten 250+ MB/s throughput with smaller buffers. Also, `BufferedInputStream` uses a buffer of 8KiB and performs reasonably.

Reply:
Testing with a buffer of 32KiB takes my laptop 20-25 seconds to hash a 13 GB file with an NVMe SSD. The CPU is an i7-8650U. Can you try to reproduce this issue with the current code base - using streams rather than a thread pool likely substantially improved the throughput.

Perhaps this occurred is if you're using multiple threads to hash on a spinny disk, therefore causing disk thrashing. You can confirm if this is the case by setting the number of hashing threads to 1, and the performance improves.

If this is what's happening we can better adjust the pipeline to compute these hashes in parallel. A single thread should be reading full chunks into memory in order and passing them into a thread pool that computes the sha1hashes. This will yield much better performance because it eliminates thrashing rather than increasing the buffer size to reduce thrashing.

As a side note, it's good that I tested this because there was a bug with creating torrents with files bigger than 2GB. This is now fixed and pushed.

Reply:
Replied by email, so my comment ended up at the top level.. Here are my measurements:

```
3 files (3.86G + 64.5M + 1.5M)
Chunk size: 16KB

# of threads, buffer size, time_cache (ms), time_no_cache (ms)
8                8MB                3488        4812
8               16KB                3650       10347
4               16KB                8981       17275
8                8KB                6435       13874
8               32KB                4643        8127
8              128KB                3980        5243
8              256KB                3693        5096
8                1MB                3681        4769
16             256KB                3515        4783
```

Time measured is the duration of `verifier.verify()` call in `bt.data.DefaultDataDescriptor.buildBitfield()` when Bt is run for an existing set of torrent files.

`time_no_cache` is captured by running Bt after having dropped the filesystem caches from RAM.

`time_cache` is captured in a second run after that (without cleaning caches).

My immediate thoughts:

- Given that the default number of hashing threads is 1, default buffer size should be significantly larger than the chunk size. I very much agree that this might not be true and/or a good thing in general, but it seems to be a fact of life in the context of current code, at least in some cases.
- Buffer size should be configurable. This way we relief ourselves of the pressure to address this immediately.
- Parallelism-related questions are worth exploring. Currently it's very difficult to say where the performance difference for a different number of hashing threads is coming from. Same for buffer sizes and other configuration options in general. No observability => making guesses and shuffling code around.
- We need metrics and performance tests. This, along with improving the logging system, is a medium-term priority.
- Both of these things merit a separate stream of work (issues, PRs, etc.):
  - Hashing improvements (threading model and buffer sizes) with regards to different storage types and torrent models.
  - Soft vs hard references. If there are indeed valid cases for buffers of different sizes, then we might need to turn this into a configuration option (or rather an intelligent dynamic behaviour) as well. On soft refs in general, we use them in other places, so it does not make much sense to address one usage instance in isolation.

Let me know your thoughts!

Reply:
Quick question before I put together a longer writeup:
> Chunk size: 16KB
This seems really small. I think 2-4MB is a much more reasonable piece size for a 4GB torrent: https://wiki.vuze.com/w/Torrent_Piece_Size

Reply:
Sorry, this was the block size. Chunk size is 1MB.

Reply:
I did a small benchmark for different buffer sizes: 16k, 32k, 64k and 16mb, and there is no difference on hashing 512mb of random data 
``` 
Benchmark                       Mode  Cnt     Score    Error  Units
HashingBenchmarks.digester_16k  avgt   10  1861.272 ± 25.412  ms/op
HashingBenchmarks.digester_32k  avgt   10  1863.558 ± 26.623  ms/op
HashingBenchmarks.digester_64k  avgt   10  1865.485 ± 25.945  ms/op
HashingBenchmarks.digester_16m  avgt   10  1899.718 ± 27.127  ms/op
```

Reply:
> Given that the default number of hashing threads is 1, default buffer size should be significantly larger than the chunk size. I very much agree that this might not be true and/or a good thing in general, but it seems to be a fact of life in the context of current code, at least in some cases.

I assume this refers to piece size, not block size. There's no advantage to having the buffer bigger than the largest piece size of any torrent that runs. This is if you look at the loop in `JavaSecurityDigest.digest()`, the loop will run the same amount of time in both cases - once.

I think (but not so strongly) that the default buffer size should be 16KiB. This is the smallest buffer that yields reasonable performance results. Anything bigger is allocating extra memory to appease OSs that don't readahead automatically.

The OS can configure readahead to prevent slowdown if torrent files are not in the filesystem cache. In linux this is viewable with `sudo blockdev --report /dev/<device>`. You can set the readahead to be 1MB by doing `sudo blockdev --setra 2048 /dev/<device>` for example. See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSPerformance.html Alternatively, the hashing threads could be increased to more than the number of CPUs to reduce I/O blocking.

Nevertheless, I will set it to 1MB for now, because breaking performance for users is never a good idea. If you'd prefer, make it larger.

> Buffer size should be configurable. This way we relief ourselves of the pressure to address this immediately.

Agreed, and written. Please review and let me know if you'd prefer the naming `hashingBufferSize` instead of `digestBufferSize`. This is also a reason to keep the current behavior.

> Currently it's very difficult to say where the performance difference for a different number of hashing threads is coming from.
> We need metrics and performance tests. This, along with improving the logging system, is a medium-term priority.

This is something with the I/O subsystem of a computer. The diversity of different hardware, and OSs will make it very difficult to test this well. And what may work well for one OS may work badly for another.

> Hashing improvements (threading model, pipelining and buffer sizes) with regards to different storage types
I don't think (or at least I'd hope) that nobody is using S3 as a filesystem to torrent things.

I have an old Dual Core Core2 System with a spinning disk I can work on this if there's interest. It's not a use case for me though, but it's interesting work and I've already figured out how the code works.

> Soft vs hard references
I think I eliminated the need for softreferences here. See: https://github.com/atomashpolskiy/bt/pull/171#discussion_r673149349
The other two usages I see for them are in CipherBufferMutator and in BufferManager.

I don't use encryption in my use case so CipherBufferMutator doesn't affect me. But the design seems wrong - assuming I am reading the code correctly, two CipherBufferMutators are constructed per peer, and then duplicated in any thread that needs this buffer. I would think at minimum that this SoftReference TL should be static to the CipherBufferMutator object, so it's at most one buffer per thread that uses it. I think that the same analysis I did for the digest verification should be done here, and it should be a static thread local. Of course, I'm biased as I've been burned before by SoftReferences. It's likely the current implementation is adequate.
 
Regarding the buffer manager, I don't like the design, but I haven't had the time to fully understand it or consider rewriting it. Allocating large direct buffers can easily leak OS memory and cause an OS OOM condition. SoftReferences won't help here, because the native allocated memory does not increase Java heap pressure causing the heap to free the buffer. I may look into this more in the future when I have some time.

I'll try and see if I can make some sense of the test results later. It's weird that a 16 KiB buffer is so much faster than a 32 KiB buffer (perhaps we're blowing the CPU's L1 Cache?), and that whatever OS you're using isn't smart enough to readahead in the files. Could you provide a bit more info on your computer (CPU/Memory/OS)?

Thank you for the detailed response and time put into this review (:

Reply:
> I assume this refers to piece size, not block size.

Yep, but the point still stands in the sense that 16KB buffer is likely to be too small, as my tests show. But they also show that starting with 256KB the performance gains start diminishing and there are no gains beyond 1MB. So yeah, that seems like a sweet spot for my particular configuration, and I don't mind making it the new default.

> I think (but not so strongly) that the default buffer size should be 16KiB. This is the smallest buffer that yields reasonable performance results. Anything bigger is allocating extra memory to appease OSs that don't readahead automatically.

This (and the rationale that follows) is interesting, but I personally think it's a bit incorrect to look at the problem from this angle. As a torrent user, I'm first and foremost concerned about download speed, then everything else. The simplest and by far the most prevalent use case is downloading a single torrent, which includes resuming an unfinished download (hence, hashing). This use case should work as fast as possible on any reasonable modern configuration out-of-the-box. If it uses 8*8 MB more RAM, so be it. It's awesome that we found that reducing this number 8-fold does not incur any performance penalties, but in general case I'd like to stress that giving up performance for any other runtime metric is not a desirable trade-off. Advanced users with more complicated use cases should have the possibility to tinker with configuration and adjust it based on measurements made on their particular OS/hardware/whatever. But this should not be considered the norm. Perhaps I'm thinking this way because I'm much more of a common user than power user myself :)

> Please review and let me know if you'd prefer the naming hashingBufferSize instead of digestBufferSize.

As you can probably say by now, I'm completely indifferent to naming. This is a user-facing configuration option though, so pick whatever makes more sense to the user in your opinion.

> This is something with the I/O subsystem of a computer. The diversity of different hardware, and OSs will make it very difficult to test this well. And what may work well for one OS may work badly for another.

Yep, that's my point (and there can also be issues in our own design as well), and it's an extra argument on why changing defaults should be approached with caution.

> I don't think (or at least I'd hope) that nobody is using S3 as a filesystem to torrent things.

AWS had an option at some point to seed the contents of a bucket. And it somewhat makes sense in general for the cloud provider to act as an always online seed. I mean, you can imagine these sort of things: [Saving torrents to S3 #121](https://github.com/atomashpolskiy/bt/issues/121).

> I don't use encryption in my use case so CipherBufferMutator doesn't affect me. But the design seems wrong - assuming I am reading the code correctly, two CipherBufferMutators are constructed per peer, and then duplicated in any thread that needs this buffer. I would think at minimum that this SoftReference TL should be static to the CipherBufferMutator object, so it's at most one buffer per thread that uses it.

Actually, messaging is single-threaded on both ends:
- inbound messages read in bt.net.DataReceivingLoop.processKey()
- outbound messages posted in bt.net.MessageDispatcher.MessageDispatchingLoop.processSupplierMap()

There are two instances of CipherBufferMutator for each peer: one for decoding the incoming messages and another for encoding the outgoing messages. Previously they were stateless (aside from the connection-specific cipher instance) and operated on the data buffer. What the current version does with thread locals is a good question. Seems like I merged this without making sure I understand the details...

> Regarding the buffer manager, I don't like the design, but I haven't had the time to fully understand it or consider rewriting it. Allocating large direct buffers can easily leak OS memory and cause an OS OOM condition.

Yep, but I/O on heap buffers is slower because it involves JVM making a copy of the data to pass it to OS. On a related note, I spent a significant amount of time in the past on implementing "direct I/O" for inbound blocks: https://github.com/atomashpolskiy/bt/pull/126 (direct in the sense that it's zero-copy all the way from JVM native buffer to storage).

> Could you provide a bit more info on your computer (CPU/Memory/OS)?

https://everymac.com/systems/apple/macbook_pro/specs/macbook-pro-core-i7-2.3-15-dual-graphics-late-2013-retina-display-specs.html

OS X 10.14.4

Reply:
> "direct I/O" for inbound blocks

Giving it a second thought and reading through my notes in that PR, maybe the outbound flow was also taken into account.

Reply:
> AWS had an option at some point to seed the contents of a bucket. 

They deprecated this: https://news.ycombinator.com/item?id=27524549

> Perhaps I'm thinking this way because I'm much more of a common user than power user myself :)

I agree - at the end of the day, the defaults should be written to satisfy the needs of the end user.

I took a compromise approach here. On Linux make the default 16 KiB, and all other OSs, make the default 1 MB. A better solution would be to explicitly hint to the OS what we're doing so it can prefetch for us. But this would require writing JNI or JNA calls that would be different for Mac/Linux, and Windows lacks a prefetch call. I changed the code to do this, and documented what native calls should be made on Mac/Linux to make this better.
This wouldn't be the first java project to do so. Hadoop has this ability, but it relies on JNI internally:
https://github.com/c9n/hadoop/blob/a315107901aebc8d0c2b9f530d191673d27922a2/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/ReadaheadPool.java#L206

Reply:
@wizzardo - out of curiosity, what OS did you test with?

Reply:
ubuntu 20.04, but my test didn't read anything from disk, just pure hashing benchmark

Reply:
OK, that won't be so helpful. The digest performance isn't likely to be affected by the size of the buffer passed in; what we're looking for is File I/O.